internal/healthchecks/healthCheck.go

```diff
@@ -17,19 +22,35 @@ func checkServerHealth(server string, healthCheckConfig config.HealthCheck) bool
 	return err == nil && res.StatusCode == 200
 }
 
-func PerformHealthChecks(config *config.Config) {
+func worker(id int, tasks <-chan HealthCheckTask) {
+	for task := range tasks {
+		if !checkServerHealth(task.Server, task.HealthCheckConfig) {
+			log.Printf("[Worker %d] Server %s failed health check, removing from pool\n", id, task.Server)
+		} else {
+			log.Printf("[Worker %d] Server %s passed health check\n", id, task.Server)
+		}
+	}
+}
+
+func PerformHealthChecks(cfg *config.Config) {
+	const numWorkers = 10
+	tasks := make(chan HealthCheckTask, 100)
+
+	for i := 0; i < numWorkers; i++ {
+		go worker(i, tasks)
+	}
+
 	for {
 		fmt.Println()
-		for _, upstream := range config.Upstreams {
-			for i := len(upstream.Servers) - 1; i >= 0; i-- {
-				server := upstream.Servers[i]
-				if !checkServerHealth(server, upstream.HealthCheck) {
-					log.Printf("Server %s failed health check, removing from pool\n", server)
-					// Remove server from slice safely.
-					upstream.Servers = append(upstream.Servers[:i], upstream.Servers[i+1:]...)
+		for _, upstream := range cfg.Upstreams {
+			for _, server := range upstream.Servers {
+				task := HealthCheckTask{
+					Server:            server,
+					HealthCheckConfig: upstream.HealthCheck,
 				}
+				tasks <- task
 			}
+			time.Sleep(upstream.HealthCheck.Interval)
```

Comment:
really think about this... is this doing what you think it should be doing?

Reply:
try to think about how to write tests for this part so you can validate the following:

- that is checks all the upstreams - ensuring the interval provided in the config is honoured.
- and that one upstreams healthcheck interval does not affect the others.

You can to some extend validate these by adding more upstreams in the config, printing the timestamps and checking the time differences manually. But you need to figure out to how actually write this as a test suite.